src/common_subset.rs

```diff
@@ -116,57 +113,95 @@ impl<NodeUid: Clone + Debug + Display + Eq + Hash + Ord> CommonSubset<NodeUid> {
         &mut self,
         message: Input<NodeUid>,
-    ) -> Result<VecDeque<Output<NodeUid>>, Error> {
+    ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
         match message {
             Input::Broadcast(uid, bmessage) => {
                 let mut instance_result = None;
-                let input_result = {
+                let input_result: Result<VecDeque<Output<NodeUid>>, Error> = {
                     if let Some(broadcast_instance) = self.broadcast_instances.get(&uid) {
                         broadcast_instance
                             .handle_broadcast_message(&uid, bmessage)
-                            .map(|(value, queue)| {
-                                instance_result = value;
+                            .map(|(opt_value, queue)| {
+                                instance_result = opt_value;
                                 queue.into_iter().map(Output::Broadcast).collect()
                             })
                             .map_err(Error::from)
                     } else {
                         Err(Error::NoSuchBroadcastInstance)
                     }
                 };
-                if instance_result.is_some() {
-                    self.on_broadcast_result(&uid)?;
+                let mut opt_message: Option<AgreementMessage> = None;
+                if let Some(value) = instance_result {
+                    self.broadcast_results.insert(uid.clone(), value);
+                    opt_message = self.on_broadcast_result(&uid)?;
                 }
-                input_result
+                input_result.map(|mut queue| {
+                    if let Some(agreement_message) = opt_message {
+                        // Append the message to agreement nodes to the common output queue.
+                        queue.push_back(Output::Agreement(agreement_message))
+                    }
+                    (None, queue)
+                })
             }
-            Input::Agreement(_uid, _message) => {
-                // FIXME: send the message to the Agreement instance and
-                // conditionally call `on_agreement_output`
 
-                Err(Error::NotImplemented)
+            Input::Agreement(uid, amessage) => {
+                // The result defaults to error.
+                let mut result = Err(Error::NoSuchAgreementInstance);
+
+                if let Some(mut agreement_instance) = self.agreement_instances.get_mut(&uid) {
+                    // Optional output of agreement and outgoing agreement
+                    // messages to remote nodes.
+                    result = if agreement_instance.terminated() {
+                        // This instance has terminated and does not accept input.
+                        Ok((None, VecDeque::new()))
+                    } else {
+                        // Send the message to the agreement instance.
+                        agreement_instance
+                            .on_input(uid.clone(), &amessage)
```

Comment:
That shouldn't be the same `uid` as above, should it? If I understand correctly:

* `self.agreement_instances.get_mut(&uid)` is the instance that decides whether the element proposed by node `uid` should be part of the common subset in the end.
* The parameter of `Agreement::on_input` should be the `uid` of the node that sent this particular message.